calyx/src/backend/rtl/gen.rs

```diff
@@ -21,278 +22,252 @@ impl Backend for RtlBackend {
-//         .append(RcDoc::line())
-//         .append(instances(c))
-//         .append(RcDoc::line())
-//         .append(RcDoc::line())
-//         .append(RcDoc::text("endmodule"));
-//     pretty_print(doc)
-// }
+    fn emit(ctx: &context::Context) -> Result<(), errors::Error> {
+        let prog: ast::NamespaceDef = ctx.clone().into();
 
-// //==========================================
-// //        Component I/O Functions
-// //==========================================
+        // build Vec of tuples first so that `comps` lifetime is longer than `docs` lifetime
+        let comps: Vec<(&ast::ComponentDef, component::Component)> = prog
+            .components
+            .iter()
+            .map(|cd| (cd, ctx.get_component(&cd.name).unwrap()))
+            .collect();
 
-// /**
-//  * Returns a string with the list of all of a component's i/o pins
-//  */
-// #[allow(unused)]
-// pub fn component_io(c: &ComponentDef) -> RcDoc<'_> {
-//     let mut inputs = c
-//         .signature
-//         .inputs
-//         .iter()
-//         .map(|pd| in_port(pd.width, &pd.name));
-//     let mut outputs = c
-//         .signature
-//         .outputs
-//         .iter()
-//         .map(|pd| out_port(pd.width, &pd.name));
-//     RcDoc::line().append(RcDoc::intersperse(
-//         inputs.chain(outputs),
-//         RcDoc::text(",").append(RcDoc::line()),
-//     ))
-// }
+        let docs = comps.iter().map(|(cd, comp)| cd.doc(&comp));
+        display(D::intersperse(docs, D::line()));
+        Ok(())
+    }
+}
 
-// pub fn in_port(width: u64, name: &str) -> RcDoc<'_> {
-//     RcDoc::text("input")
-//         .append(RcDoc::space())
-//         .append(RcDoc::text("logic"))
-//         .append(RcDoc::space())
-//         .append(bit_width(width))
-//         .append(RcDoc::text(name))
-// }
+impl Emitable for ast::ComponentDef {
+    fn doc<'a>(&'a self, comp: &'a component::Component) -> D<'a, ColorSpec> {
+        D::text("// Component Signature")
+            .append(D::line())
+            .append(colors::define(D::text("module")))
+            .append(D::space())
+            .append(&self.name)
+            .append(D::line())
+            .append(parens(
+                D::line()
+                    .append(self.signature.doc(&comp))
+                    .nest(4)
+                    .append(D::line()),
+            ))
+            .append(";")
+            .append(D::line())
+            .append(D::line())
+            .append("// Wire declarations")
+            .append(D::line())
+            .append(wire_declarations(&comp))
+            .append(D::line())
+            .append(D::line())
+            .append("// Subcomponent Instances")
+            .append(D::line())
+            .append(subcomponent_instances(&comp))
+            .append(D::line())
+            .append(colors::define(D::text("endmodule")))
+            .append(D::space())
+            .append(format!("// end {}", &self.name))

```

Comment:
Is there a reason to use sequences of `append` instead of appending one giant vector?

Reply:
It’s not a vector. I’m appending to the RcDoc type